server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -500,4 +399,249 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
 
         return this;
     }
+
+    public static abstract class Values extends AbstractCollection implements Writeable {
+
+        private static Values readFrom(StreamInput in) throws IOException {
+            if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+                return in.readOptionalWriteable(ValuesAfterV8::new);
+            } else {
+                List<?> list = (List<?>)in.readGenericValue();
+                return list == null ? null : new ValuesBeforeV8(list);
+            }
+        }
+
+        private static void writeTo(StreamOutput out, Values values) throws IOException {
+            if (out.getVersion().onOrAfter(Version.V_8_0_0)) {
+                out.writeOptionalWriteable(values);
+            } else {
+                if (values == null) {
+                    out.writeGenericValue(null);
+                } else {
+                    values.writeTo(out);
+                }
+            }
+        }
+
+        protected static BytesRef serialize(Iterable<?> values, boolean convert) {
+            List<?> list;
+            if (values instanceof List<?>) {
+                list = (List<?>) values;
+            } else {
+                ArrayList<Object> arrayList = new ArrayList<>();
+                for (Object o : values) {
+                    arrayList.add(o);
+                }
+                list = arrayList;
+            }
+            try (BytesStreamOutput output = new BytesStreamOutput()) {
+                if (convert) {
+                    list = list.stream().map(AbstractQueryBuilder::maybeConvertToBytesRef).collect(Collectors.toList());
+                }
+                output.writeGenericValue(list);
+                return output.bytes().toBytesRef();
```

Comment:
Drive-by comment: this call to `BytesReference#toBytesRef()` is typically going to allocate a whole new large `byte[]` just to copy all the pages of the stream into it. That seems to be an unnecessary expense, I don't see a need for the result of this method to be contiguous in memory, so let's use `BytesReference` rather than `BytesRef` to avoid this copying.

Reply:
@DaveCTurner Thanks for the review! fixed in [b2b9afd](https://github.com/elastic/elasticsearch/pull/67223/commits/b2b9afd5e81a9921abef3cbf25aff1d9083d87b0)